Constructor documentation (#639)
KDAB/cxx-qt: book/src/concepts/constructor.md

* the Rust struct can't have a `default` implementation as it needs certain arguments for construction.

To facilitate these use-cases CXX-Qt provides the [`Constructor` trait][constructor-trait].

## Implementing & Declaring a Constructor
In order for CXX-Qt to generate a custom constructor, the `qobject::T` must implement the [`cxx_qt::Constructor` trait][constructor-trait].
Additionally, the constructor must be declared within the [cxx_qt::bridge](./bridge.md).

This declaration uses Rust's `impl ... for` syntax, but with a few special rules:
* The implementation block must be empty
* If any of the associated types in the constructor are not `()`, they
    must be listed using `Type=(...)` in the constructor generics.

Example:
```rust,ignore,noplayground
#[cxx_qt::bridge]
mod qobject {
{{#include ../../../examples/qml_features/rust/src/signals.rs:book_signals_struct}}

{{#include ../../../examples/qml_features/rust/src/signals.rs:book_constructor_decl}}
}
// Don't forget to actually implement the trait **outside** the bridge!
```
[Full example](https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/signals.rs)

## Routing arguments of the Constructor

> A C++ constructor is more complex then a Rust struct initialization.

Author: Done

For this it uses three associated types as well as one generic type:
* Constructor generic `Arguments` - argument list that the constructor takes.
* `type BaseArguments` - argument list that is passed to the base class constructor.
* `type NewArguments` - argument list used to construct the Rust struct using the `new` function.
* `type InitializeArguments` - argument list that is passed to the code run within the constructor (the `initialize` function).

These types must all be tuple types and can therefore support an arbitrary list of arguments.

The `route_arguments` function takes the Arguments provided to the constructor and distributes them to the three different types in the above list.
Doing this in Rust allows calling any base class constructor and controlling movement, copying or other operations on the arguments.
Every constructor defined using `cxx_qt::Constructor` will execute this function once, before anything else in the Constructor happens and then distributes the resultant values to the appropriate steps of the QObject construction.

> As actual construction takes place in C++, all types used must be compatible with CXX as they must be able to pass the FFI barrier.

Author: Done

## Constructing the Rust struct with arguments
If the construction of the inner Rust struct requires some arguments, the `new` function of the Constructor can be given arguments by using the `NewArguments` associated type.

Example:
```rust,ignore
#[cxx_qt::bridge]
mod qobject {
    // ...

{{#include ../../../examples/qml_features/rust/src/properties.rs:book_constructor_new_decl}}
}

{{#include ../../../examples/qml_features/rust/src/properties.rs:book_constructor_new_impl}}
```

## Initializing the QObject
In addition to running code before constructing the inner Rust struct, it may be useful to run code from the context of the QObject itself (i.e. inside the Constructor implementation).

The `initialize` function can be used to run code inside a constructor.
It is given a pinned mutable self reference to the QObject and the list of `InitializeArguments`.

### Using the `Initialize` trait

> As QML uses the default constructor, for most QML types only the `initialize` part of the constructor is of interest, as everything else is defaulted.

Author: Done

Every constructor defined using `cxx_qt::Constructor` will execute this function once, before anything else in the Constructor happens and then distributes the resultant values to the appropriate steps of the QObject construction.

As actual construction takes place in C++, all types used must be compatible with CXX as they must be able to pass the FFI barrier.

### Default constructor
To implement a default constructor in CXX-Qt (the constructor used by QML) a good default would be to just leave everything as `()`.
The `route_arguments` implementation would then look like this:
```rust,ignore
impl cxx_qt::Constructor<()> for qobject::T {
    type NewArguments = ();
    type InitializeArguments = ();
    type BaseArguments = ();

    fn route_arguments(_arguments: ()) -> (
        Self::NewArguments,
        Self::BaseArguments,
        Self::InitializeArguments,
    ) {
        // Nothing to route
        ((),(),())
    }
}
```

## Constructing the Rust struct with arguments
If the construction of the inner Rust struct requires some arguments, the `new` function of the Constructor can be given arguments by using the `NewArguments` associated type.

Example:
```rust,ignore

> impl cxx_qt::Constructor<(i32,)> for qobject::T {

Reviewer: do we not have an examples where these can be pulled from ? As we've been trying to remove blocks of untested/built code from the book / docs

Author: Currently not, as qml_features uses QML, which always default-constructs anyway.
I can add a dummy constructor to properties to make sure it compiles though.

Reviewer: :thinking: just write a comment next to it saying the dummy one has no purpose other than for documentation so folks don't think it's required :-) 

Was just trying to ensure we don't add more code blocks that could be tested.

<!-- 
SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>

SPDX-License-Identifier: MIT OR Apache-2.0
-->

# Defining a custom C++/QML constructor

By default, CXX-Qt will generate a constructor that takes a single `QObject` pointer as optional argument (usually the QObject parent).

It then:
1. Calls the base class constructor with the optional parent pointer.

> 2. uses [Rusts default trait](https://doc.rust-lang.org/std/default/trait.Default.html) to construct the inner Rust struct.

Reviewer 2: ```suggestion
2. uses [Rust's default trait](https://doc.rust-lang.org/std/default/trait.Default.html) to construct the inner Rust struct.
```

Author: Done

> For most cases this is good enough, however there are a few cases in which this will not be sufficient.
> 
> Examples are if...

Reviewer 2: ```suggestion
For most cases this is good enough, however there are a few cases in which this will not be sufficient, for example:
```

Author: Done

> * the base class constructor expects arguments other than the parent pointer.

Reviewer 2: ```suggestion
* the base class constructor expects arguments other than the parent QObject pointer, or a different type than QObject is used for the parent (for example QQuickItem's constructor takes a QQuickItem pointer)
```

Author: Done

* the QObject needs to run code after it is initialized (e.g. open a connection to a server, etc.).
* the Rust struct can't have a `default` implementation as it needs certain arguments for construction.

To facilitate these use-cases CXX-Qt provides the [`Constructor` trait][constructor-trait].

## Implementing & Declaring a Constructor
In order for CXX-Qt to generate a custom constructor, the `qobject::T` must implement the [`cxx_qt::Constructor` trait][constructor-trait].
Additionally, the constructor must be declared within the [cxx_qt::bridge](./bridge.md).

> This declaration uses Rusts `impl ... for` syntax, but with a few special rules:

Reviewer 2: ```suggestion
This declaration uses Rust's `impl ... for` syntax, but with a few special rules:
```

* The implementation block must be empty
* If any of the associated types in the constructor are not `()`, they
    must be listed using `Type=(...)` in the constructor generics.

Example:
```rust,ignore,noplayground
#[cxx_qt::bridge]
mod qobject {
{{#include ../../../examples/qml_features/rust/src/signals.rs:book_signals_struct}}

{{#include ../../../examples/qml_features/rust/src/signals.rs:book_constructor_decl}}
}
```

## Routing arguments of the Constructor
A C++ constructor is more complex then a Rust struct initialization.

Like a normal function, it takes a list of arguments.
However, before the actual code inside the constructor is executed, C++ ensures the following:
* The base class is constructed
* All members of the struct are initialized.

Both of these steps may use any of the given arguments.

cxx_qt::Constructor allows controlling how the arguments are passed around in the constructor using the `route_arguments` function.

> For this it uses three Associated types as well as one generic type:

Reviewer 2: ```suggestion
For this it uses three associated types as well as one generic type:
```

Author: Done
